### PR TITLE
Drop support for OpenSSL <1.1.1

### PR DESCRIPTION
**Pull request by sethmlarson**
Closes https://github.com/urllib3/urllib3/issues/2168

From the original issue, decided to not to implement these:

- Didn't remove detection for SSLContext.minimum_version because it requires Python 3.8
- Didn't remove SSLKEYLOGFILE detection because it requires Python 3.8
- Didn't remove post_handshake_auth detection because it requires Python 3.8

### CODE REVIEW

**Comment by pquentin** [comment]
OK the CI passes now! I had to make changes for docs, requests, botocore and coverage. I believe the main unresolved question is now LibreSSL in the issue.

**Comment by sethmlarson** [comment]
I wonder about boto and `DEFAULT_CIPHERS`. Instead of removing it should we carry this list around as vestigal with a `DeprecationWarning` on the module level access of it?

**Comment by nateprewitt** [comment]
We could potentially internalize the cipher list in Botocore if you don't want to maintain this into 2.0. I think the reason they were imported currently was the assumption that urllib3 would do a better job of making sure they were up to date.

**Comment by pquentin** [comment]
My thinking was that if botocore upgrades to urllib3 2.0 and thus enforces OpenSSL 1.1.1+ it would make sense to trust the default OpenSSL cipher list in botocore too. (urllib3 1.26.x currently uses the existing OpenSSL 1.1.1 cipher list on all OpenSSL versions.) This is of course up to you @nateprewitt. You don't really have to decide now, I think either way the value of the botocore downstream tests is the same?

**Comment by sethmlarson** [comment]
I'm defending the usage of `fmt: off` here as this is what Black produced by default:

```python
    if not ssl.OPENSSL_VERSION.startswith("OpenSSL ") or ssl.OPENSSL_VERSION_INFO < (
        1,
        1,
        1,
    ):  # Defensive:
        raise ImportError(
            "urllib3 v2.0 only supports OpenSSL 1.1.1+, currently "
            f"the 'ssl' module is compiled with {ssl.OPENSSL_VERSION}. "
            "See: https://github.com/urllib3/urllib3/issues/2168"
        )
```

File: test/test_ssl.py [hunk 1]
```diff
@@ -172,24 +133,18 @@ def test_create_urllib3_context_pha(
 
         assert context.post_handshake_auth == expected_pha
 
-    @pytest.mark.parametrize("use_default_sslcontext_ciphers", [True, False])
     def test_create_urllib3_context_default_ciphers(
-        self, monkeypatch: pytest.MonkeyPatch, use_default_sslcontext_ciphers: bool
+        self,
+        monkeypatch: pytest.MonkeyPatch,
```
Comment (pquentin) [inline test/test_ssl.py:138]: nit: removing the trailing comma will use one less line

**Review by pquentin — CHANGES_REQUESTED** [review]
Thanks! We're nearly there. I left a bunch of nits. I also think we can remove two more things:

* According to https://docs.python.org/3/library/ssl.html#ssl.SSLContext.maximum_version `minimum_version` and `maximum_version` are Python 3.7+, so any detection can be removed? (Not sure where it's done)
* We can also remove `hasattr(context, "load_default_certs")` in two places as `load_default_certs` is available in Python 3.4+: https://docs.python.org/3/library/ssl.html#ssl.SSLContext.load_default_certs

**Comment by sethmlarson** [comment]
@pquentin Thanks! Addressed your review comments.

For the `load_default_certs` we need to continue checking for availability as both the pyOpenSSL and SecureTransport SSLContext implementations don't support the method.

We don't do any detection for `minimum_version` and `maximum_version`, original implementation wouldn't attempt to use the feature unless the value was specified.

File: test/test_ssl.py [hunk 2]
```diff
@@ -42,67 +42,21 @@ def test_is_ipaddress_true(self, addr: Union[bytes, str]) -> None:
...
-    @pytest.mark.parametrize(
-        ["ciphers", "expected_ciphers"],
-        [
-            (None, ssl_.DEFAULT_CIPHERS),
-            ("ECDH+AESGCM:ECDH+CHACHA20", "ECDH+AESGCM:ECDH+CHACHA20"),
-        ],
-    )
     def test_create_urllib3_context_set_ciphers(
         self,
         monkeypatch: pytest.MonkeyPatch,
```
Comment (pquentin) [inline test/test_ssl.py:47]: nit: new trailing comma :)

**Review by pquentin — APPROVED** [review]
Thanks, this looks great to me!

> For the `load_default_certs` we need to continue checking for availability as both the pyOpenSSL and SecureTransport SSLContext implementations don't support the method.

Oh I see. Maybe we should mention that to know when to remove it?

**Review by pquentin — APPROVED** [review]
LGTM! Thanks for bearing with me 😅

**Comment by martinwunderlich** [comment]
Just FYI, I have run into some issues related to this removal of DEFAULT_CIPHERS. I am trying to run a python lambda function on AWS with some dependencies that in turn depend on package request 2.30.0 which depens on urllib3 2.0.2. 
I guess I'll somehow have to downgrade urllib or update botocore. 
There is a related issue here in the requests project: https://github.com/psf/requests/issues/6438

```
{
    "errorMessage": "Unable to import module 'mycode': cannot import name 'DEFAULT_CIPHERS' from 'urllib3.util.ssl_' (/opt/python/urllib3/util/ssl_.py)",
    "errorType": "Runtime.ImportModuleError",
    "requestId": "id",
    "stackTrace": []
}
```

**Comment by martinwunderlich** [comment]
@pquentin Thanks for the feedback! I've downgraded urllib to <2 (which resolved to 1.26.15) and now it works.

**Pull request merged**